torch/_utils.py

```diff
@@ -774,31 +781,48 @@ def _get_device_index(
     device of the supported runtime platform if :attr:`optional` is ``True``.
     i.e., the current default CUDA device will be returned if CUDA runtime is supported.
     """
+    if isinstance(device, int):
+        return device
     if isinstance(device, str):
         device = torch.device(device)
-    device_idx: Optional[int] = None
     if isinstance(device, torch.device):
+        if allow_cpu and device.type == "cpu":
+            return -1
+        if torch.jit.is_scripting():
+            if device.type != "cuda":
+                raise ValueError(
+                    f"Expected a cuda{' or cpu' if allow_cpu else ''} device, but got: {device}"
+                )
+        else:
+            # _get_available_device_type is an eager API which is not scriptable.
+            available_device_type = _get_available_device_type()
```

Comment:
I will let the usual reviewers look at this but this is very much not in line with the doc.
It would be great to make these actually device generic (including in the name and intent) rather than nesting if statements.

Reply:
The parameter `device` can be `torch.device`, `int`, `str`, `None` and device guard class, eg `torch.cuda.device` or `torch.xpu.device`. And `_get_device_index` is usually used in eager mode and JIT mode. So I extend if statements to handle each situation. I intend to refine this code further and would appreciate any additional input or suggestions for improvement.

Reply:
I think https://github.com/pytorch/pytorch/issues/108046 and the other issues linked there are relevant on how to extend this.

Reply:
Thanks for your input.